### ISSUE TITLE
PDF file Upload

### ISSUE THREAD

**Issue by DataDrug** [jonalxh/Flask-Admin-Dashboard#3]
Hello,

I'm learning python the worst way possible (I believe), and upon using your dashboard as a template I'm  stuck as many doubt's, such as: How to make direct chat work, connecting data to widget's etc, however, I'd just woul'd like to understand how could I add a new column on the user list so a CV could be imported. Would you be kind enough to explain me or provide me a snippet so I could study it? At this point the flas-admin documentation is a little confusing for me.

Thank you very much.

**Comment by jonalxh** [comment]
Hi @DataDrug.

I don't recommend Flask-Admin as a learning point, because of its use requires you to have some experience about Python and the way it works.

Your doubts are a little more focused on front-end side (chat and widgets functionality), if you want to make it work i recommend you to use Flask-Admin standalone views, which let you have views without database dependence, my project contains one example in the menu item called Custom View, give it a look. But for chatting maybe yo need to use SocketIO or whatever realtime plugin.

For PDF uploading give a look to the flask-admin file upload documentation, [here](https://flask-admin.readthedocs.io/en/latest/advanced/#managing-files-folders) is a little example about how to manage the FileAdmin functionality in a new view, I've not tested it but sounds very simple. By the other side, if you need to add a file upload input, they explain how to make it work [here](https://flask-admin.readthedocs.io/en/latest/api/mod_form_upload/). 

Lately I'm full of work and it is not possible for me to try this kind of things, in a free time I'll try. For the moment please try with the info provided, if you succeed It would be a great help if you share us your improved code in a pull request.

I hope I've helped a little.

**Comment by DataDrug** [comment]
Hi again,

I've looked into both links previously, but they are really confusing (they may be simple for someone who understands python, but not for me at the moment), i'll take a look at it this weekend and share the news.

Thank you.

**Comment by DataDrug** [comment]
So, this was what I managed to do so far.

My main objective was to have a database of people who applied for "my company" with 3 fields:

Name, Job Applied, Curriculum.

Now I just need to understand the download routine, but I'll tackle other issues and doubts I have before that.

I won't commit because this might be full of errors or bad practices but it worked for me anyways.

If anyone ever has this doubt, here is the code.


` 

# Create directory for file fields to use
file_path = op.join(op.dirname(__file__), 'uploadedfiles')
try:
    os.mkdir(file_path)
except OSError:
    pass


# Create models
class Curriculums(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    Name = db.Column(db.Unicode(64))
    Job = db.Column(db.Unicode(64))
    CV = db.Column(db.Unicode(128))

    def __unicode__(self):
        return self.name

# Delete hooks for models, delete files if models are getting deleted
@listens_for(Curriculums, 'after_delete')
def del_file(mapper, connection, target):
    if target.path:
        try:
            os.remove(op.join(uploadedfiles, target.path))
        except OSError:
            # Don't care if was not deleted because it does not exist
            pass

# Administrative views
class FileView(sqla.ModelView):

    def is_accessible(self):
        if not current_user.is_active or not current_user.is_authenticated:
            return False

        if current_user.has_role('superuser'):
            return True

        return False

    def _handle_view(self, name, **kwargs):
        """
        Override builtin _handle_view in order to redirect users when a view is not accessible.
        """
        if not self.is_accessible():
            if current_user.is_authenticated:
                # permission denied
                abort(403)
            else:
                # login
                return redirect(url_for('security.login', next=request.url))


    can_edit = True
    edit_modal = True
    create_modal = True
    can_export = True
    can_view_details = True
    details_modal = True
    column_searchable_list = ['Job']

    # Override form field to use Flask-Admin FileUploadField
    form_overrides = {
        'CV': form.FileUploadField
    }

    # Pass additional parameters to 'path' to FileUploadField constructor
    form_args = {
        'CV': {
            'label': 'CV',
            'base_path': file_path,
            'allow_overwrite': True
        }
    } `

**Comment by jonalxh** [comment]
It sounds very nice. Later i'll take some time to implement it in the repo.
I'm glad you managed to do it.

**Comment by DataDrug** [comment]
Hi again,

I was wondering, how could I grab the number of entries on the Curriculums class and serve it on the dashboard? Is it something hard to do or really simple and I can't google the right question for a right answer?

Another issue I've ran into is:

1) I don't know how to specify file mimetypes for uploading
2) Files with characters like é, ó, ç, ã, í, can't be uploaded (I've placed the UTF-8 thingy at the top of app.py but not even in other fields it works, I always get a syntax error.
3) After I deleted the entry, the files still remain on the 'uploadedfiles' folder.

Anyways, I've been learning more 12h of python, so even tho I'm dead tired, it was a good day.

Thank's again for all your help.
